gluetool/glue.py

```diff
@@ -1557,160 +1629,176 @@ def has_module_class(item):
 
-                self.modules[mname] = {
-                    'class': cls,
-                    'description': cls.description,
-                    'group': group
-                }
+        for root, _, files in os.walk(dirpath):
+            for filename in sorted(files):
+                if not filename.endswith('.py'):
+                    continue
 
-                loaded_modules.append((group, cls))
+                # A group of the module is defined by the directories it lies in under the ``dirpath``.
+                if root == dirpath:
+                    group_name = ''
+                else:
+                    group_name = root.replace(dirpath + os.sep, '')
 
-            # if name is a list, add more aliases to the same module
-            if isinstance(module_names, (list, tuple)):
-                for mname in module_names:
-                    add_module(mname, cls)
+                pm_name = '{}.{}.{}'.format(
+                    pm_prefix,
+                    group_name.replace(os.sep, '.'),
+                    os.path.splitext(filename)[0]
+                )
 
-            else:
-                add_module(module_names, cls)
+                self._discover_gm_in_file(registry, os.path.join(root, filename), pm_name, group_name)
 
-        return loaded_modules
+    def _discover_gm_in_entry_point(self, entry_point, registry):
+        # type: (str, Dict[str, DiscoveredModule]) -> None
 
-    def _load_module_path(self, ppath):
-        # type: (str) -> None
+        self.debug('discovering modules in entry point {}'.format(entry_point))
 
+        for ep_entry in pkg_resources.iter_entry_points(entry_point):
+            klass = ep_entry.load()
+
+            self._register_module(registry, getattr(klass, 'group', ''), klass, ep_entry.dist.location)
+
+    def discover_modules(self, entry_points=None, paths=None):
+        # type: (Optional[List[str]], Optional[List[str]]) -> Dict[str, DiscoveredModule]
         """
-        Search and load ``gluetool`` modules from a directory.
+        Discover and load all accessible modules.
+
+        Two sources are examined:
 
-        In essence, it scans every file with ``.py`` suffix, and searches for
-        classes derived from :py:class:`gluetool.glue.Module`.
+        1. entry points, handled by setuptools, to which Python packages can attach ``gluetool`` modules they provide,
+        2. directory trees.
 
-        :param str ppath: directory to search for `gluetool` modules.
+        :param list(str) entry_points: list of entry point names to which ``gluetool`` modules are attached.
+            If not set, entry points set byt the configuration (``--module-entry-point`` option) are used.
+        :param list(str) paths: list of directories to search for ``gluetool`` modules. If not set, paths set by
+            the configuration (``--module-path`` option) are used.
+        :rtype: dict(str, DiscoveredModule)
```

Comment:
`ModuleRegistryType` ?

Reply:
Should be fixed.

Reply:
Will be fixed in a separate patch that should cover use of Mypy types in Sphinx directives in general.